distributed/comm/tcp.py

```diff
@@ -0,0 +1,254 @@
+import errno
+import logging
+import socket
+import struct
+import sys
+
+from toolz import first
+
+from tornado import gen
+from tornado.iostream import IOStream, StreamClosedError
+from tornado.tcpclient import TCPClient
+from tornado.tcpserver import TCPServer
+
+from .. import config
+from .core import connectors, listeners, Comm, Listener, CommClosedError
+from .utils import to_frames, from_frames, parse_host_port, unparse_host_port
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_total_physical_memory():
+    try:
+        import psutil
+        return psutil.virtual_memory().total / 2
+    except ImportError:
+        return 2e9
+
+
+MAX_BUFFER_SIZE = get_total_physical_memory()
+
+
+def set_tcp_timeout(stream):
+    """
+    Set kernel-level TCP timeout on the stream.
+    """
+    if stream.closed():
+        return
+
+    timeout = int(config.get('tcp-timeout', 30))
+
+    sock = stream.socket
+
+    # Default (unsettable) value on Windows
+    # https://msdn.microsoft.com/en-us/library/windows/desktop/dd877220(v=vs.85).aspx
+    nprobes = 10
+    assert timeout >= nprobes + 1, "Timeout too low"
+
+    idle = max(2, timeout // 4)
+    interval = max(1, (timeout - idle) // nprobes)
+    idle = timeout - interval * nprobes
+    assert idle > 0
+
+    try:
+        if sys.platform.startswith("win"):
+            logger.debug("Setting TCP keepalive: idle=%d, interval=%d",
+                         idle, interval)
+            sock.ioctl(socket.SIO_KEEPALIVE_VALS, (1, idle * 1000, interval * 1000))
+        else:
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
+            try:
+                TCP_KEEPIDLE = socket.TCP_KEEPIDLE
+                TCP_KEEPINTVL = socket.TCP_KEEPINTVL
+                TCP_KEEPCNT = socket.TCP_KEEPCNT
+            except AttributeError:
+                if sys.platform == "darwin":
+                    TCP_KEEPIDLE = 0x10  # (named "TCP_KEEPALIVE" in C)
+                    TCP_KEEPINTVL = 0x101
+                    TCP_KEEPCNT = 0x102
+                else:
+                    TCP_KEEPIDLE = None
+
+            if TCP_KEEPIDLE is not None:
+                logger.debug("Setting TCP keepalive: nprobes=%d, idle=%d, interval=%d",
+                             nprobes, idle, interval)
+                sock.setsockopt(socket.SOL_TCP, TCP_KEEPCNT, nprobes)
+                sock.setsockopt(socket.SOL_TCP, TCP_KEEPIDLE, idle)
+                sock.setsockopt(socket.SOL_TCP, TCP_KEEPINTVL, interval)
+
+        if sys.platform.startswith("linux"):
+            logger.debug("Setting TCP user timeout: %d ms",
+                         timeout * 1000)
+            TCP_USER_TIMEOUT = 18  # since Linux 2.6.37
+            sock.setsockopt(socket.SOL_TCP, TCP_USER_TIMEOUT, timeout * 1000)
+    except EnvironmentError as e:
+        logger.warn("Could not set timeout on TCP stream: %s", e)
+
+
+class TCP(Comm):
+
+    def __init__(self, stream, peer_addr, deserialize=True):
```

Comment:
I got the impression that some comms were deserialize-always and others deserialize-never, though I've also kept the deserialize argument on `read()` for now. Which approach do you think is better?

If you're talking about the more general approach of pickling and unpickling in read() and write(), yes, I think it's desirable: imagine we write a new transport usable between threads of the same process, in this case we won't want to pickle anything, simple pass objects around in a queue.